org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/outline/ModelBuilder.scala

```diff
@@ -0,0 +1,242 @@
+package org.scalaide.ui.internal.editor.outline
+
+import scala.tools.nsc.Settings
+import scala.tools.nsc.interactive.CompilerControl
+import scala.tools.nsc.interactive.Global
+import scala.tools.nsc.reporters.ConsoleReporter
+import scala.reflect.internal.util.BatchSourceFile
+import scala.reflect.internal.util.SourceFile
+import scala.tools.nsc.interactive.Response
+import org.scalaide.core.compiler.IScalaPresentationCompiler
+import org.scalaide.logging.HasLogger
+import scala.collection.mutable.MutableList
+import scala.reflect.internal.util.OffsetPosition
+
+/**
+ * ModelBuilder is the "heart" of Outline View. It maps scala source into a tree of nodes (model). Each node
+ * represent a scala entity, such as Class, Trait, Method and so on.
+ * This tree is rendered by TreeView later. Parsed tree is used to build the model. Which means not all
+ * type information is available.
+ * The model is built on the reconciler thread. Every time new model is generated it overwrites the previous
+ * one. There are two reasons why the model is mutable.
+ *
+ * 1. Despite of good model builder performance, applying model to TreeView could be expansive. For large file
+ * tens of thousands nodes are generated. Each node is translated to UI widget and all widget manipulation
+ * is done on UI thread, which may slow down the application.
+ *
+ * 2. The SWT TreeView has a state- user may expand/close some nodes. If we replace model, that state is lost.
+ *
+ * Using mutable model allows to update only nodes are modified.
+ */
+object ModelBuilder extends HasLogger {
+  private val tuple = """scala.Tuple[0-9]+""".r
+  private val func = """_root_.scala.Function[0-9]+""".r
+  def buildTree(comp: IScalaPresentationCompiler, src: SourceFile): RootNode = {
+    import comp._
+    import scala.reflect.internal.Flags._
+    def setPos(n: Node, t: Tree) = {
+      if (t.pos.isDefined) {
+        n.end = t.pos.end
+        n.start = t.pos.start
+      }
+    }
+
+    def updateTree(parent: ContainerNode, t: Tree): Unit = {
+      def renderTuple(sb: StringBuilder, args: List[Tree]) = {
+        sb.append("(")
+        args.foreach(a => { renderType(sb, a); sb.append(", ") })
+        if (!args.isEmpty)
+          sb.setLength(sb.length - 2)
+        sb.append(")")
+      }
+      def renderFunc(sb: StringBuilder, args: List[Tree]): Unit = {
+        args match {
+          case a :: b :: tail =>
+            renderType(sb, a)
+            sb.append(", ")
+            renderFunc(sb, b :: tail)
+          case a :: Nil =>
+            sb.setLength(sb.length - 2)
+            sb.append(") => ")
+            renderType(sb, a)
+          case Nil =>
+        }
+      }
+      def renderATT(sb: StringBuilder, tt: Tree, tpt: Tree, args: List[Tree], dp: Boolean): Unit = {
+        tpt.toString match {
```

Comment:
I think this would be *much* cleaner by matching against the actual type structure, than calling `toString` and using regular expressions to de-structure. All the structured information is already in the tree.